Include exception causes

xiongtx opened the pull request:
Fix #7.

HB [supports causes](http://blog.honeybadger.io/honeybadger-now-supports-causes-nested-exceptions/). See the [updated documentation](http://docs.honeybadger.io/guides/api.html#sample-payload) on including causes in the payload.

Example:

![image](https://user-images.githubusercontent.com/1624090/29583842-a54c7830-8736-11e7-910f-f2f174d0222c.png)

camdez commented:
@xiongtx Thank you for your work on this! I actually added support for *a* cause exactly a week ago and wasn't sure how to handle nested causes gracefully. Looks like you have the answer!

Give me just a bit to look at over and I'll get it merged in!

camdez commented on src/honeybadger/core.clj:
```diff
@@ -28,6 +28,13 @@
+  "Return all exceptions in the exception chain."
+  [e]
+  (->> e
+       (iterate #(try (.getCause %) (catch NullPointerException _)))
```
Is there a need for this that can't be obviated by simply checking if the value is `nil`? (I'd rather not cause and catch exceptions in non-exceptional circumstances.)

xiongtx replied:
:white_check_mark:

camdez commented on src/honeybadger/core.clj:
```diff
@@ -37,15 +44,23 @@
   (->> st st/parse-trace-elems (map format-stacktrace-elem)))
 
 (defprotocol Notifiable
+  (error-map-no-causes [this])
```
Do you see a compelling reason to expand the protocol? This seems like just a concern of `Throwable`'s implementation of the protocol, not the protocol itself.

xiongtx replied:
:white_check_mark:

camdez commented on src/honeybadger/core.clj:
```diff
@@ -37,15 +44,23 @@
+     :backtrace (format-stacktrace (.getStackTrace this))})
+  (error-map [this]
+    (let [exs (rest (ex-chain this))]
+      (merge (error-map-no-causes this)
```
This could be simplified to an `assoc`.

xiongtx replied:
:white_check_mark:

xiongtx commented:
Addressed comments :ok_hand:.

> Looks like you have the answer!

The folks at @honeybadger-io kindly updated the documentation so I could figure out what's going on :clap:.

camdez commented:
@xiongtx Looking good! Let me do a little testing and then I'll get it merged / cut a new release. (Also gonna rebase to `dev`. I've only been merging to `master` for release.)

> The folks at @honeybadger-io kindly updated the documentation so I could figure out what's going on 👏.

Thanks, kind @honeybadger-io folks! 🎉

camdez commented:
Merged! (Though apparently not in a way that made GitHub properly realize it was merged. :/ )

I've pushed a snapshot (pre-release) version to Clojars: https://clojars.org/camdez/honeybadger/versions/0.3.1-SNAPSHOT

I'm going to run that in one of my own production environments for a day or two just to make sure we're solid, and then I'll ship a proper 0.4.0 release.

Thanks again for your hard work! 🥂

xiongtx commented:
> I'm going to run that in one of my own production environments for a day or two 

Sounds good, but it'd be nice if there were better tests. We can then have greater confidence that changes have't introduced regressions.

One thing I'd like to see is replacing `schema` w/ `spec` and specing out a [notice](https://github.com/honeybadger-io/honeybadger-ruby/blob/master/lib/honeybadger/notice.rb).

camdez commented:
@xiongtx completely agree on both. PRs welcome. 😜 Or I’ll eventually find the time. 

FWIW I did add a quick test for these changes, but the project obviously needs more.
